### PR TITLE
Refactored logs and added log-to-json feature (removed the old log-to-file feature), added logging test utils.

### LINKED ISSUE #365: Implementing structured logs in JSON format
This is mostly for the tests. The idea is to output all the logs in a JSON format, so it will be much much easier to compare them (we won't have to deal with the log_widht, etc). 

This is fairly easy to implement. See https://docs.python.org/3/howto/logging-cookbook.html#implementing-structured-logging

### PR DESCRIPTION
**Pull request by huguesdevimeux**
fixes #365

 ## List of Changes
- Refactored the log creation in `logger.py`. Before, it was using  `_run_config()` (a private function, in addition), and was using the config parser to fetch the logger configuration in the .cfg, although it was already done by `config.py` Now, logger.py contains two functions, `set_rich_logger` and `set_file_logger` which are used to set up the handlers of the logs directly in `config.py`. cc @leotrs  @naveen521kk
- Added log to file feature. Before, it was done by saving the RIch console text, which for me was problematic for two reasons : It's a pure hell to debug as one has to deal with all the whitespaces and formats stuff. The second reason was that it didn't use the `FileHandler` logging handler. Now, it uses `FIleHandler` with a `Formatter`, so the logs will be exported in a json format so it will be easily readable and parseable ! cc @Aathish04 . Note : It does not change the output in `sys.out` 
- Testings utils to test logs : Added `@logs_comparison` decorator to test logs (same as `@video_comparison`). Example of usage : 
```python
@logs_comparison(
    "BasicSceneLoggingTest.txt", os.path.join("logs", "SquareToCircle.log")
)
def test_logging_to_file(tmp_path, python_version):
    path_basic_scene = os.path.join("tests", "test_logging", "basic_scenes.py")
    # os.makedirs(path_output, exist_ok=True)
    command = [
        python_version,
        "-m",
        "manim",
        path_basic_scene,
        "SquareToCircle",
        "-l",
        "--log_to_file",
        "--media_dir",
        str(tmp_path),
    ]
    out, err, exitcode = capture(command)
    assert exitcode == 0, err.decode()
```

## Motivation
THis wasn't right IMO. 

## Further comment : 
Example of a log saved : 
```JSON
{"levelname": "INFO", "module": "config", "message": "Log file wil be saved in <>"}
{"levelname": "ERROR", "module": "__main__", "message": "\n   Tezst is not in the script\n"}
{"levelname": "INFO", "module": "scene", "message": "Animation 0 : Using cached data (hash : <>)"}
{"levelname": "INFO", "module": "scene_file_writer", "message": "\nFile ready at <>\n"}
{"levelname": "INFO", "module": "scene", "message": "Rendered SquareToCircle\nPlayed 1 animations"}

...
```
Note the "<>". This is to avoid having some user-dependent stuff in the logs, in particular during tests.
**IMPORTANT !**
We don't use any regex to hide this kind of informatons. TO hide some in a logs, simply do : 
```python
logger.info("test, %(info)s", {"info": topsecretinfotohide}).
```
Basically, everything contained in the dict passed as arg will be hidden
## Testing Status
Created a new test, test_logging.py. The main idea of this PR was not to add new tests but rather test-logs tools.  

## Acknowledgement
- [ x] I have read the [Contributing Guidelines](https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP))

### CODE REVIEW

**Comment by huguesdevimeux** [comment]
Review at will guys !

File: docs/source/changelog.rst [hunk 1]
```diff
@@ -86,6 +86,7 @@ Of interest to developers
 #. Added autogenerated documentation with sphinx and autodoc/autosummary [WIP]
 #. Made manim internally use relative imports
 #. Since the introduction of the :code:`TexTemplate` class, the files :code:`tex_template.tex` and :code:`ctex_template.tex` have been removed
+#. Added logging tests tools.
```
Comment (leotrs) [inline docs/source/changelog.rst:89]: Added ability to save logs in json

File: manim/config/config_utils.py [hunk 2]
```diff
@@ -69,9 +69,15 @@ def _parse_file_writer_config(config_parser, args):
...
-    for str_opt in ["media_dir", "log_dir"]:
+    for str_opt in ["media_dir"]:
         attr = getattr(args, str_opt)
         fw_config[str_opt] = os.path.relpath(default[str_opt]) if attr is None else attr
+    attr = getattr(args, "log_dir")
+    fw_config["log_dir"] = (
+        os.path.join(fw_config["media_dir"], default["log_dir"])
+        if attr is None
+        else attr
+    )
```
Comment (leotrs) [inline manim/config/config_utils.py:80]: Question: I don't understand why this is put here and the default.cfg was changed from `%(media_dir)s/logs` to just `logs`. These two seem to have the same result.

File: manim/config/default.cfg [hunk 3]
```diff
@@ -174,7 +174,7 @@ log_message =
 log_path = dim
 log_width = -1
 log_height = -1
-log_timestamps = True
+log_timestamps = False
```
Comment (leotrs) [inline manim/config/default.cfg:177]: Question: is this necessary? I kinda like being able to tell how long the rendering took.

File: tests/test_logging/test_logging.py [hunk 4]
```diff
@@ -4,65 +4,26 @@
 import pytest
```
Comment (leotrs) [inline tests/test_logging/test_logging.py:4]: (I'll put this here because I don't know where else to put it. GH doesn't allow comments on deleted/empty files.)

Question: Why is the `test_logging/__init__.py` necessary?

File: tests/test_logging/basic_scenes.py [hunk 5]
```diff
@@ -5,15 +5,7 @@
 
 class SquareToCircle(Scene):
```
Comment (leotrs) [inline tests/test_logging/basic_scenes.py:6]: Nitpick: `SquareAndCircle` maybe?

File: manim/scene/scene.py [hunk 6]
```diff
@@ -65,7 +65,10 @@ def construct(self):
     def __init__(self, **kwargs):
         Container.__init__(self, **kwargs)
         self.camera = self.camera_class(**camera_config)
-        self.file_writer = SceneFileWriter(self, **file_writer_config,)
+        self.file_writer = SceneFileWriter(
+            self,
+            **file_writer_config,
+        )
```
Comment (leotrs) [inline manim/scene/scene.py:71]: Comment: This is only happening because of the trailing comma. I believe if you delete the trailing comma, black will just shut up about this line.

File: manim/scene/scene.py [hunk 7]
```diff
@@ -195,7 +198,10 @@ def update_frame(  # TODO Description in Docstring
         if file_writer_config["skip_animations"] and not ignore_skipping:
             return
         if mobjects is None:
-            mobjects = list_update(self.mobjects, self.foreground_mobjects,)
+            mobjects = list_update(
+                self.mobjects,
+                self.foreground_mobjects,
+            )
```
Comment (leotrs) [inline manim/scene/scene.py:204]: Same comment RE trailing comma.

File: manim/scene/scene_file_writer.py [hunk 8]
```diff
@@ -608,21 +616,12 @@ def flush_cache_directory(self):
         for f in cached_partial_movies:
             os.remove(f)
         logger.info(
-            f"Cache flushed. {len(cached_partial_movies)} file(s) deleted in {self.partial_movie_directory}."
+            f"Cache flushed. {len(cached_partial_movies)} file(s) deleted in %(par_dir)s.",
+            {"par_dir": self.partial_movie_directory},
```
Comment (leotrs) [inline manim/scene/scene_file_writer.py:620]: Same thing about trailing comma (x6 in this file)

File: tests/test_logging/test_logging.py [hunk 9]
```diff
@@ -4,65 +4,26 @@
...
-            path_output,
-            "-v",
-            "DEBUG",
-            "--config_file",
-            os.path.join("tests", "test_logging", "testloggingconfig.cfg"),
-        ]
-    )
-    out, err, exitcode = capture(command, use_shell=True)
-    log_file_path = os.path.join(path_output, "logs", "SquareToCircle.log")
+    # os.makedirs(path_output, exist_ok=True)
```
Comment (leotrs) [inline tests/test_logging/test_logging.py:16]: Nitpick: ;)

File: tests/utils/logging_tester.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+
+def logs_comparison(control_data_file, log_path_from_media_dir):
+    """Decorator used for any test that needs to check logs.
+
+    Parameters
+    ----------
+    control_data_file : :class:`str`
+        Name of the control data file, i.e. .log that will be compared to the outputed logs.
+        .. warning:: You don't have to pass the path here.
```
Comment (leotrs) [inline tests/utils/logging_tester.py:47]: Issue: I don't understand what this warning is telling me. If I don't have to pass the path, then what do I need to pass?

File: tests/utils/logging_tester.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+    log_path_from_media_dir : :class:`str`
+        The path of the .log generated, from the media dir. Example: /logs/Square.log.
+    """
+
+    def decorator(f):
+        @wraps(f)
+        def wrapper(*args, **kwargs):
+            # NOTE : Every args goes seemingly in kwargs instead of args; this is perhaps Pytest.
```
Comment (leotrs) [inline tests/utils/logging_tester.py:56]: This doesn't seem right. Can you provide an example of this?

File: tests/utils/logging_tester.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+            result = f(*args, **kwargs)
+            tmp_path = kwargs["tmp_path"]
+            tests_directory = os.path.dirname(
+                os.path.dirname(os.path.abspath(__file__))
+            )
+            path_control_data = os.path.join(
+                tests_directory, "control_data", "logs_data", control_data_file
+            )
```
Comment (leotrs) [inline tests/utils/logging_tester.py:64]: Nitpick: one variable is `path_control_data` while the other one is `control_data_file`. Why not `control_data_path` and `control_data_file`?

File: tests/utils/logging_tester.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
+from functools import wraps
+import os
+import json
+
+
+def _check_logs(reference_logfile, generated_logfile):
+    reference_logs = open(reference_logfile, "r").readlines()
+    generated_logs = open(generated_logfile, "r").readlines()
```
Comment (leotrs) [inline tests/utils/logging_tester.py:8]: Issue :warning: : these files are not being closed

File: tests/utils/logging_tester.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+    if len(reference_logs) != len(generated_logs):
+        msg_assert = ""
+        if len(reference_logs) > len(generated_logs):
+            diff = len(reference_logs) - len(generated_logs)
```
Comment (leotrs) [inline tests/utils/logging_tester.py:13]: Nitpick (non-blocking): lines 13 and 18 are duplicated. They could be moved outside of the `if`

File: tests/utils/logging_tester.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+            msg_assert += f"Logs generated are SHORTER than the expected logs. There are {diff} extra logs.\n"
+            msg_assert += "Last log of the generated log is : \n"
+            msg_assert += generated_logs[-1]
+        else:
+            diff = len(generated_logs) - len(reference_logs)
+            msg_assert += f"Logs generated are LONGER than the expected logs.\n There are {diff} extra logs :\n"
+            for log in generated_logs[len(reference_logs) :]:
+                msg_assert += log
```
Comment (leotrs) [inline tests/utils/logging_tester.py:21]: Comment. I'm not loving the asymmetry here. Why output only the last line in one case but all lines in the other?

Comment (leotrs) [reply tests/utils/logging_tester.py:21]: Also, why not use a regular diff? I understand they can be long and hard to read.... are there any other reasons?

File: tests/utils/logging_tester.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        assert 0, msg_assert
+
+    for index, (ref, gen) in enumerate(zip(reference_logs, generated_logs)):
```
Comment (leotrs) [inline tests/utils/logging_tester.py:24]:
```suggestion
    for index, ref, gen in zip(itertools.count(), reference_logs, generated_logs):
```

Comment (leotrs) [reply tests/utils/logging_tester.py:24]: ^ nitpick

File: tests/utils/logging_tester.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        ref_log = json.loads(ref)
+        gen_log = json.loads(gen)
+        diff_keys = [
+            d1[0] for d1, d2 in zip(ref_log.items(), gen_log.items()) if d1[1] != d2[1]
+        ]
+        # \n does not work in f-strings.
+        newline = "\n"
+        assert (
+            len(diff_keys) == 0
+        ), f"Lgos don't match at {index} st log. : \n{newline.join([f'In {key} field, got -> {gen_log[key]}. {newline}Expected : -> {ref_log[key]}.' for key in diff_keys])}"
```
Comment (leotrs) [inline tests/utils/logging_tester.py:37]: Comment: I know it's black and not you, but I _hate_ this indentation.

Also typos "**Lgos** don't match at {} **st** log."

Comment (leotrs) [reply tests/utils/logging_tester.py:37]: Also it'd be nice if these two lines lined up one on top of the other:

`In {key} field god -> `
`Expected ->`

You can try to do line them up by introducing `\t` chars in the second one

File: manim/config/config.py [hunk 18]
```diff
@@ -150,3 +152,14 @@ def _parse_config(config_parser, args):
         _init_dirs(file_writer_config)
 config = _parse_config(config_parser, args)
 camera_config = config
+
+# Set the different loggers
+set_rich_logger(config_parser["logger"])
+if file_writer_config["log_to_file"]:
+    # IMPORTANT note about file name : The log file name will be the scene_name get from the args (contained in file_writer_config). So it can differ from the real name of the scene.
```
Comment (leotrs) [inline manim/config/config.py:159]: Question: what if the user doesn't pass a scene name? `manim -pl scene.py`. I'd use the `input_file` to create the name.

Also typo: get -> gotten

**Review by leotrs — CHANGES_REQUESTED** [review]
One general comment: right now this seems to implement a either: a rich logger to the console, or BOTH a rich logger and a JSON logger to a file.

I think it'd be nicer to have the option either format (rich, json), to either output (console, file). I understand that's not what this PR is about, but it might be a refactor needed in the future.

File: manim/config/default.cfg (hunk 3 above)
Comment (huguesdevimeux) [reply manim/config/default.cfg:177]: This is not necessary, no. But I thought it would be appreciated as timestamps in the rich formatted logs are very, very annoying and ugly. 

I can change it back if you think it should be so.

File: tests/test_logging/test_logging.py (hunk 4 above)
Comment (huguesdevimeux) [reply tests/test_logging/test_logging.py:4]: It's for pytest. Basically, if I don't put this there, I can't use `utils` 

see https://docs.pytest.org/en/stable/goodpractices.html#test-package-name

File: tests/test_logging/basic_scenes.py (hunk 5 above)
Comment (huguesdevimeux) [reply tests/test_logging/basic_scenes.py:6]: I don't think so, as this is the same as `SquareToCircle` scene ;)

File: manim/scene/scene_file_writer.py (hunk 8 above)
Comment (huguesdevimeux) [reply manim/scene/scene_file_writer.py:620]: Yes, this is super annoying. I will see what I can do

File: tests/test_logging/test_logging.py (hunk 9 above)
Comment (huguesdevimeux) [reply tests/test_logging/test_logging.py:16]:
```suggestion
```

(just do quick commit)

File: tests/utils/logging_tester.py (hunk 10 above)
Comment (huguesdevimeux) [reply tests/utils/logging_tester.py:47]: Just the file name of the .log (for example, `SquareToCircleWithLFlag.log`. I think I will add an example.

File: tests/utils/logging_tester.py (hunk 11 above)
Comment (huguesdevimeux) [reply tests/utils/logging_tester.py:56]: An example of what ?

File: tests/utils/logging_tester.py (hunk 12 above)
Comment (huguesdevimeux) [reply tests/utils/logging_tester.py:64]: I don't underswtand your comment. 

One is the file name and the other is the entire path.

File: tests/utils/logging_tester.py (hunk 13 above)
Comment (huguesdevimeux) [reply tests/utils/logging_tester.py:8]: Oooooow yes thank you so much. I will correct this !

File: tests/utils/logging_tester.py (hunk 15 above)
Comment (huguesdevimeux) [reply tests/utils/logging_tester.py:21]: Yes, the main reason is that they _are_ often long and hard to read. Furthermore, the whole point of having json formatted logs is be able to do pretty differences display 👍🏻 

Concerning the asymetry, I've been thunking of this and I think it's better. Like, when generated logs are shorter than what expected, it's better to see what is the last log generated than showing all the logs that should have been there, as these can be accessed in the expected logs.

File: manim/scene/scene_file_writer.py (hunk 8 above)
Comment (leotrs) [reply manim/scene/scene_file_writer.py:620]: I think they will all go away if you just remove the trailing comma and run black again.

File: manim/config/default.cfg (hunk 3 above)
Comment (leotrs) [reply manim/config/default.cfg:177]: I have no strong opinion about showing logs or not.

I do have a strong opinion about changing `default.cfg` without first reaching consensus, because this affects every single user.

File: tests/test_logging/test_logging.py (hunk 4 above)
Comment (leotrs) [reply tests/test_logging/test_logging.py:4]: got it, thanks

File: tests/utils/logging_tester.py (hunk 15 above)
Comment (leotrs) [reply tests/utils/logging_tester.py:21]: I'm not convinced that one is necessarily better than the other in one case or another. But you have more experience in debugging test logs so you should have final sya.

File: manim/config/default.cfg (hunk 3 above)
Comment (huguesdevimeux) [reply manim/config/default.cfg:177]: Alright. I’ll change this back

**Comment by huguesdevimeux** [comment]
> One general comment: right now this seems to implement a either: a rich logger to the console, or BOTH a rich logger and a JSON logger to a file.
> 
> I think it'd be nicer to have the option either format (rich, json), to either output (console, file). I understand that's not what this PR is about, but it might be a refactor needed in the future.

To answer : I'm not sure ti understand your question. If you're wondering how to only save logs in file and nothing in the console, then it's completly possible by using `-v`, since the `FIleHandler` used set is own log level. 

Thanks for the review !

File: manim/config/config.py (hunk 18 above)
Comment (huguesdevimeux) [reply manim/config/config.py:159]: Yes, good idea, but the advantage of this is when there is several scenes in one file and the user wants to keep logs from each scenes, without having only one log file for the whole scenes file.

But I think this is a better idea. Should we use the file name instead of the scene name passed as arg ?

Comment (leotrs) [reply manim/config/config.py:159]: I'm fairly neutral about this. In my experience, it is rare to have more than one Scene in the same file in real use cases. It also depends on the size of the log files.

OTOH, if we ever implement parallel rendering of scenes, we may want to keep the logs separated. So let's keep it like this :shrug:

**Comment by leotrs** [comment]
> To answer : I'm not sure ti understand your question.

My question is: do we have a way of implementing each of the following?
+ output rich logs to the console and nothing to file
+ output JSON logs to the console and nothing to file
+ output rich logs to the console and rich to file
+ output JSON logs to the console and rich to file
+ output rich logs to the console and JSON to file
+ output JSON logs to the console and JSON to file
+ output nothing to the console and rich to file
+ output nothing to the console and JSON to file

File: manim/config/config.py (hunk 18 above)
Comment (huguesdevimeux) [reply manim/config/config.py:159]: Done  :D

I also _hate_ this formatting.

**Comment by huguesdevimeux** [comment]
@leotrs Well no. This PR removed the option to output rich console to a file as it is now replaced. 
It is not possible to output the JSON to the console as well. 
Basically, the options available are  : 
- output rich logs to the console and nothing to file
- output rich logs to the console and JSON to file
- output nothing to the console and JSON to file

**Comment by leotrs** [comment]
Thanks. In the future, I think it'd be worth it to refactor this in order to be able to have all those combinations.

**Comment by huguesdevimeux** [comment]
> Thanks. In the future, I think it'd be worth it to refactor this in order to be able to have all those combinations.

Maybe. I am not convinced of the usefulness tho, but wht not. 

By the way, I addressed all your requests. Feel free to check this :D

File: tests/utils/logging_tester.py (hunk 13 above)
Comment (PgBiel) [reply tests/utils/logging_tester.py:8]: I recommend using `with`

File: tests/utils/logging_tester.py (hunk 12 above)
Comment (PgBiel) [reply tests/utils/logging_tester.py:63]: he means the names should follow a better pattern

File: tests/utils/logging_tester.py [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    """Decorator used for any test that needs to check logs.
+
+    Parameters
+    ----------
+    control_data_file : :class:`str`
+        Name of the control data file, i.e. .log that will be compared to the outputed logs.
+        .. warning:: You don't have to pass the path here.
+
+    log_path_from_media_dir : :class:`str`
+        The path of the .log generated, from the media dir. Example: /logs/Square.log.
```
Comment (PgBiel) [inline tests/utils/logging_tester.py:49]: should likely add a Returns here, with Callable

File: tests/utils/logging_tester.py (hunk 12 above)
Comment (huguesdevimeux) [reply tests/utils/logging_tester.py:63]: Ow yea, forgive my stupidity. I wil change this rn

**Comment by huguesdevimeux** [comment]
I adressed everything, I think.

File: manim/config/config_utils.py (hunk 2 above)
Comment (huguesdevimeux) [reply manim/config/config_utils.py:80]: IT does not work otherwise for a strange reason, and I think I just discovered a new issue. 
When passing `--media_dir` as flag, `log_dir` (and I think tex(t)_dir as well), their paths are not updated with the media_dir passed as flag. 

To confirm tho

Comment (behackl) [reply manim/config/config_utils.py:80]: I noticed this as well, although for images and videos, this is the reason why I had to set their directories separately in https://github.com/ManimCommunity/manim/pull/402/files#diff-0d9eb0ee2926ea2e8371ae8f9842a28dR155-R157

Comment (leotrs) [reply manim/config/config_utils.py:80]: Oh I see yeah, this is a bug of the stupid config system. Let's leave this change in since the config system is being totally revamped soon either way.

File: tests/utils/logging_tester.py [hunk 20]
```diff
@@ -30,11 +31,12 @@ def _check_logs(reference_logfile, generated_logfile):
...
             d1[0] for d1, d2 in zip(ref_log.items(), gen_log.items()) if d1[1] != d2[1]
         ]
-        # \n does not work in f-strings.
+        # \n and \t don't not work in f-strings.
         newline = "\n"
+        tab = "\t"
         assert (
             len(diff_keys) == 0
-        ), f"Lgos don't match at {index} st log. : \n{newline.join([f'In {key} field, got -> {gen_log[key]}. {newline}Expected : -> {ref_log[key]}.' for key in diff_keys])}"
+        ), f"Lgos don't match at {index} st log. : \n{newline.join([f'In {key} field, got -> {newline}{tab}{repr(gen_log[key])}. {newline}Expected : -> {newline}{tab}{repr(ref_log[key])}.' for key in diff_keys])}"
```
Comment (leotrs) [inline tests/utils/logging_tester.py:39]: typo Lgos

Comment (leotrs) [reply tests/utils/logging_tester.py:39]: typo st

**Review by leotrs — CHANGES_REQUESTED** [review]

File: tests/utils/logging_tester.py (hunk 20 above)
Comment (huguesdevimeux) [reply tests/utils/logging_tester.py]: Yes, sorry, I completly forget that. Fixed now

**Review by behackl — CHANGES_REQUESTED** [review]
Overall this LGTM!

Could you fix the merge conflict?

File: manim/config/logger.py [hunk 21]
```diff
@@ -49,51 +64,55 @@ def parse_theme(fp):
...
-    console = Console(record=True)
-    printf(
-        "[logging.level.warning]No cfg file found, creating one in "
-        + successfully_read_files[0]
-        + " [/logging.level.warning]"
+    return customTheme
+
+
+def set_rich_logger(config_logger):
+    """Will set the RichHandler of the logger.
```
[the remark this answers is not in the document]
Comment (huguesdevimeux) [reply manim/config/logger.py:71]: Hmm I _think_ it's because logger.py location has been changed by a recent config refactor. Any thought about this @leotrs ?

[the remark this answers is not in the document]
Comment (leotrs) [reply manim/config/logger.py:73]: No, it's because sphinx doesn't like that I renamed the config module during import as `import manim.config as _config`. I'm working on a fix that will change that (but I have 0 time this week so).

Let's merge this as is, or just remove the logger from documentation for now.

[the remark this answers is not in the document]
Comment (behackl) [reply manim/config/logger.py:73]: +1, it was not meant as a blocking comment.

**Comment by leotrs** [comment]
There are now 48 files in the diff and I don't have time to look for your actual changes :+1: 

Next time when resolving conflicts, I think rebasing your branch is better than just merging from master.

**Comment by huguesdevimeux** [comment]
@leotrs  I see only 16 files in the diff :/

tests should be fixed now. 

And, sorry for the mess.

**Review by leotrs — APPROVED** [review]

**Review by behackl — APPROVED** [review]
The failing `black` test does not need to be fixed; this is caused by something that is already fixed in master -- and thus, merging will solve it.

Other than that, this LGTM! :-)

File: manim/config/logger.py [hunk 22]
```diff
@@ -18,24 +18,41 @@
 from rich.theme import Theme
 from rich import print as printf
 from rich import errors, color
+import json
+import copy
 
-from .config_utils import _run_config
 
+class JSONFormatter(logging.Formatter):
+    """Subclass of `:class:`logging.Formatter`, to build our own format of the logs (JSON)."""
```
Comment (PgBiel) [inline manim/config/logger.py:26]:
```suggestion
    """Subclass of :class:`logging.Formatter`, to build our own format of the logs (JSON)."""
```

File: manim/config/logger.py [hunk 23]
```diff
@@ -49,51 +66,54 @@ def parse_theme(fp):
...
-        + " [/logging.level.warning]"
+    return customTheme
+
+
+def set_rich_logger(config_logger, verbosity):
+    """Will set the RichHandler of the logger.
+
+    Parameter
+    ----------
+    config_logger :class:
```
Comment (PgBiel) [inline manim/config/logger.py:77]:
```py
    Parameters
    ----------
    config_logger : :class:`??? which class is it???`
```

File: tests/utils/logging_tester.py [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+        Name of the control data file, i.e. .log that will be compared to the outputed logs.
+        .. warning:: You don't have to pass the path here.
+        .. example:: "SquareToCircleWithLFlag.log"
+
+    log_path_from_media_dir : :class:`str`
+        The path of the .log generated, from the media dir. Example: /logs/Square.log.
+
+    Returns
+    -------
+    Callable[[Any], Any]
```
Comment (PgBiel) [inline tests/utils/logging_tester.py:58]: is this a function taking one argument of any type? If you mean a function which may have any signature (parameters), consider:
```suggestion
    Callable[[...], Any]
```
